Adding Readme and updated webhook files (#2)
catchpoint/Integrations.GoogleCloudMonitoring: Stackdriver-Webhook/README.md

# Catchpoint and Google's Cloud Monitoring & Logging

## Introduction

Cloud Monitoring with Catchpoint provides visibility into cloud services and the applications they depend on. Cloud services depend on many different services, from Nginx to Elasticsearch to Cassandra. With Cloud Monitoring, Operations teams can proactively monitor their infrastructure with data collected from application instrumentation, probes, events, and metadata. With integrations to collaboration tools, alerts can be sent via Slack, PagerDuty, and others. And with advanced charting and dashboards, analysts can isolate issues and recover quickly.

This integration uses the [Cloud Monitoring](https://cloud.google.com/monitoring) REST API to ingest raw data from Catchpoint Test Data webhook.

### Catchpoint Tests

#### Supported Metrics

> 1. **Connect** The time to establish a connection with a specific URL. Reported in milliseconds.

Reviewer: ```suggestion
1. **Connect**: The time to establish a connection with a specific URL. Reported in milliseconds.
```

> 1. **DNS** The time to resolve the primary URL. Reported in milliseconds.

Reviewer: ```suggestion
1. **DNS**: The time to resolve the primary URL. Reported in milliseconds.
```

> 1. **Content Load** The time to load all components, from the first byte to the last byte, from the provided URL. Reported in milliseconds.

Reviewer: ```suggestion
1. **Content Load**: The time to load all components, from the first byte to the last byte, from the provided URL. Reported in milliseconds.
```

> 1. **Document Complete** The time to render the full webpage. Reported in milliseconds.

Reviewer: ```suggestion
1. **Document Complete**: The time to render the full webpage. Reported in milliseconds.
```

> 1. **Round Trip Time** The time it takes for a network request to go from a starting point to a destination and back again to the starting point.

Reviewer: ```suggestion
1. **Round Trip Time**: Time difference between the client initiating a request and the client receiving a response. Reported in milliseconds.
```

> 1.  **Packet Loss** occurs when one or more packets of data travelling across a computer network fail to reach their destination.

Reviewer: ```suggestion
1.  **Packet Loss**: Percentage of total packet loss.
```

#### Supported tests and available metrics

> 1. **Web Test** Connect, DNS, Content Load, Document Complete.

Reviewer: ```suggestion
1. **Web Test**: Connect, DNS, Content Load, Document Complete.
```

> 1. **Transaction Test** Connect, DNS, Content Load, Document Complete.

Reviewer: ```suggestion
1. **Transaction Test**: Connect, DNS, Content Load, Document Complete.
```

> 1. **API Test** Connect, DNS, Content Load

Reviewer: ```suggestion
1. **API Test**: Connect, DNS, Content Load.
```

> 1. **Trace route Test** :  Packet loss , round trip time, number of hops.

Reviewer: ```suggestion
1. **Traceroute Test** :  Packet Loss , Round Trip Time, Number of Hops.
```

> 1. **Ping Test** :  Packet Loss, Round Trip Time.

Reviewer: ```suggestion
1. **Ping Test**:  Packet Loss, Round Trip Time.
```

> 1. **DNS Test** :  Response times.

Reviewer: Please specify the metrics collected.
```suggestion
1. **DNS Test**:  <metrics>.
```

Author: For the DNS test, we are capturing only response time.
Added **Response time** as the metric.

<img width="960" alt="dns_metric" src="https://user-images.githubusercontent.com/68921257/92390868-53d32d80-f139-11ea-91cd-1b953d70dfee.png">

## Installation and Configuration

 ### Google Cloud setup:
 
 #### Enabling the Monitoring API
 
_Note: Cloud Monitoring requires a Google Cloud project that is associated with a Workspace._

1. Navigate to your Google Cloud [project dashboard](https://console.cloud.google.com/apis/dashboard) to access your API information.

1. Find and select `Cloud Monitoring API` to ensure APIs are enabled.
_Note: If the API is already enabled, the message_ **API enabled** _will be displayed. If there is no message, click on the_ **Enable** _button._

#### Install cloud SDK on your local machine

1. Download the [Cloud SDK installer.](https://dl.google.com/dl/cloudsdk/channels/rapid/GoogleCloudSDKInstaller.exe)

1. Launch the installer and follow the prompts. The installer is signed by Google LLC. Cloud SDK requires Python. Supported versions are 3.5 to 3.7, and 2.7.9 or higher. The installer will install all necessary dependencies, including the needed Python version.

1. After installation has completed, After installation has completed, select `Start Cloud SDK Shell` option.

_Note : The installer starts a terminal window and runs the_ `gcloud init` _command._

> #### Enabling cloud functions.

Reviewer: ```suggestion
#### Enabling Google Cloud functions.
```

1. Go to the project selector page

1. Make sure that billing is enabled for your Google Cloud project.   [Learn how to confirm billing is enabled for your project](https://cloud.google.com/billing/docs/how-to/modify-project).

1. Enable the [Cloud Functions and Cloud Pub/Sub APIs.](https://console.cloud.google.com/flows/enableapi?apiid=cloudfunctions,pubsub&redirect=https://cloud.google.com/functions/docs/tutorials/pubsub)

1. Install and initialize the Cloud SDK.

> 1. Update gcloud components:\

Reviewer: ```suggestion
1. Update gcloud components:
```

`$ gcloud components update`

#### Set up the Google Cloud Monitoring Repository locally

> 1. Clone this repository into a working directory\

Reviewer: ```suggestion
1. Clone this repository into a working directory
```

`$ git clone https://github.com/catchpoint/Integrations.GoogleCloudMonitoring.git`

1. In the `.env` file fromCatchpoint-Stackdriver-Webhook directory , update GoogleProjectId

> #### Deploying cloud functions.

Reviewer: ```suggestion
#### Deploying Google Cloud functions.
```

> Index.js has two functions called catchpointPublish and catchpointSubscribe.

Reviewer: No need for this line. "Index.js has two functions called catchpointPublish and catchpointSubscribe."

Open Google Cloud SDK Shell and navigate to the directory where the NodeJS scripts was extracted.

 `$ cd <path to extracted directory/Integrations.GoogleCloudMonitoring/Stackdriver-Webhook/>;`
 

>  1. Deploy publish function:\

Reviewer: ```suggestion
 1. Deploy publish function:
```

    `$ gcloud functions deploy catchpointPublish --trigger-http --runtime nodejs10 --timeout=180 --trigger-http --allow-unauthenticated`
        
    Copy the URL once the deployment is successful. This will be webhook URL which will be added in Catchpoint portal.
    

>  1. Deploy Subscribe function:\

Reviewer: ```suggestion
 1. Deploy Subscribe function:
```

    `$ gcloud functions deploy catchpointSubscribe --trigger-topic catchpoint-webhook --timeout=180 --runtime nodejs10 --allow-unauthenticated`
    
####  Set up the Catchpoint Test data webhook.

Add the copied URL to Catchpoint.

> 1. In catchpoint,from Settings go to [API page](https://portal.catchpoint.com/ui/Content/Administration/ApiDetail.aspx).

Reviewer: ```suggestion
1. In Catchpoint, from Settings go to [API page](https://portal.catchpoint.com/ui/Content/Administration/ApiDetail.aspx).
```

1. Under Test Data Webhook add the copied url.

> 1. Select default JSON for Test-data webhook and save the changes

Reviewer: ```suggestion
1. Select default JSON for Test Data Webhook and save the changes
```

> _Note: Test data webhook should be enabled under the test properties page._

Reviewer: ```suggestion
_Note: Test Data Webhook should be enabled under the test properties page._
```

# Catchpoint and Google's Cloud Monitoring & Logging

## Introduction

Cloud Monitoring with Catchpoint provides visibility into cloud services and the applications they depend on. Cloud services depend on many different services, from Nginx to Elasticsearch to Cassandra. With Cloud Monitoring, Operations teams can proactively monitor their infrastructure with data collected from application instrumentation, probes, events, and metadata. With integrations to collaboration tools, alerts can be sent via Slack, PagerDuty, and others. And with advanced charting and dashboards, analysts can isolate issues and recover quickly.

> This integration uses the [Cloud Monitoring](https://cloud.google.com/monitoring) REST API to ingest raw data from Catchpoint Test dat webhook.

Reviewer: Typo: should be `Data`

### Catchpoint Tests

#### Supported Metrics

1. **Connect** The time to establish a connection with a specific URL. Reported in milliseconds.

1. **DNS** The time to resolve the primary URL. Reported in milliseconds.

1. **Content Load** The time to load all components, from the first byte to the last byte, from the provided URL. Reported in milliseconds.

1. **Document Complete** The time to render the full webpage. Reported in milliseconds.

> 

Reviewer: With the addition of the traceroute and ping, please add the new metrics and definitions here. Then add to the Supported tests and available metrics listing.

#### Supported tests and available metrics

1. **Web Test** Connect, DNS, Content Load, Document Complete  

1. **Transaction Test** Connect, DNS, Content Load, Document Complete

1. **API Test** Connect, DNS, Content Load

1. **Trace route Test** :  Packet loss , round trip time, number of hops.

1. **Ping Test** :  Packet Loss, Round Trip Time.

1. **DNS Test** :  Response times.

> 

Reviewer: nit: Remove extra empty line.

##  Prerequisites

- Google Cloud project
- An active Catchpoint account

## Installation and Configuration

 ### Google Cloud setup:
 
 #### Enabling the Monitoring API
 
_Note: Cloud Monitoring requires a Google Cloud project that is associated with a Workspace._

1. Navigate to your Google Cloud [project dashboard](https://console.cloud.google.com/apis/dashboard) to access your API information.

1. Find and select `Cloud Monitoring API` to ensure APIs are enabled.
_Note: If the API is already enabled, the message_ **API enabled** _will be displayed. If there is no message, click on the_ **Enable** _button._

> 

Reviewer: nit: Remove extra empty line.

#### Install cloud SDK on your local machine

1. Download the [Cloud SDK installer.](https://dl.google.com/dl/cloudsdk/channels/rapid/GoogleCloudSDKInstaller.exe)

1. Launch the installer and follow the prompts. The installer is signed by Google LLC. Cloud SDK requires Python. Supported versions are 3.5 to 3.7, and 2.7.9 or higher. The installer will install all necessary dependencies, including the needed Python version.

> 1. After installation has completed, accept the following options:\

Reviewer: ```suggestion
1. After installation has completed, select `Start Cloud SDK Shell` option.
```

    1. Start Cloud SDK Shell

>  Note : The installer starts a terminal window and runs the gcloud init command.

Reviewer: ```suggestion
    _Note : The installer starts a terminal window and runs the_ `gcloud init` _command._
```

#### Enabling cloud functions.

1. Go to the project selector page

1. Make sure that billing is enabled for your Google Cloud project.   [Learn how to confirm billing is enabled for your project](https://cloud.google.com/billing/docs/how-to/modify-project).

1. Enable the [Cloud Functions and Cloud Pub/Sub APIs.](https://console.cloud.google.com/flows/enableapi?apiid=cloudfunctions,pubsub&redirect=https://cloud.google.com/functions/docs/tutorials/pubsub)

1. Install and initialize the Cloud SDK.

> 1. Update gcloud components:

Reviewer: For consistency.

```suggestion
1. Update gcloud components:
`$ gcloud components update`
```

 ```bash
$ gcloud components update
```
#### Set up the Google Cloud Monitoring Repository locally

1. Clone this repository into a working directory\
`$ git clone https://github.com/catchpoint/Integrations.GoogleCloudMonitoring.git`

1. In the `.env` file fromCatchpoint-Stackdriver-Webhook directory , update GoogleProjectId

>     

Reviewer: nit: Remove extra empty line.

#### Deploying cloud functions.

Index.js has two functions called catchpointPublish and catchpointSubscribe.
Open Google Cloud SDK Shell and navigate to the directory where the NodeJS scripts was extracted.

>  ```bash

Reviewer: ```suggestion
 `$ cd <path to extracted directory/Integrations.GoogleCloudMonitoring/Stackdriver-Webhook/>;`
```

$ cd <path to extracted directory/Integrations.GoogleCloudMonitoring/Stackdriver-Webhook/>;
```
 1. Deploy publish function:

>     ```bash

Reviewer: ```suggestion
    `$ gcloud functions deploy catchpointPublish --trigger-http --runtime nodejs10 --timeout=180 --trigger-http --allow-unauthenticated`
```

    $ gcloud functions deploy catchpointPublish --trigger-http --runtime nodejs10 --timeout=180 --trigger-http --allow-unauthenticated
    ```
    Copy the URL once the deployment is successful. This will be webhook URL which will be added in Catchpoint portal.
 
 1. Deploy Subscribe function:

>     ```bash

Reviewer: ```suggestion
    `$ gcloud functions deploy catchpointSubscribe --trigger-topic catchpoint-webhook --timeout=180 --runtime nodejs10 --allow-unauthenticated`
```

    $ gcloud functions deploy catchpointSubscribe --trigger-topic catchpoint-webhook --timeout=180 --runtime nodejs10 --allow-unauthenticated
    ```
####  Set up the Catchpoint Test data webhook.

Add the copied URL to Catchpoint.

> 1. In catchpoint,from Settings go to [API page](https://portal.catchpoint.com/ui/Content/Administration/ApiDetail.aspx).

Reviewer: nit: Add empty line here.

> 1. Under Test Data Webhook add the copied url.

Reviewer: nit: Add empty line here.

1. Select default JSON for Test-data webhook and save the changes

> 

Reviewer: nit: Remove extra empty line.

> 

Reviewer 2: Lets remove the unwanted blank lines.

Author: Most of the blank lines were removed on the latest commits

# Catchpoint Google Stackdriver

  

> ####

Reviewer 2: is this required?

Author: changed on the latest commits

> ## **Introduction**

Reviewer 2: just '##' are enough for headings, no need for '**'.

Author: were changed on the latest commits

  

Cloud Monitoring provides visibility into the performance, uptime, and overall health of applications. It collects metrics, events, and metadata from Google Cloud, Amazon Web Services, hosted uptime probes, application instrumentation, and a variety of common application components including Cassandra, Nginx, Apache Web Server, Elasticsearch, and many others. Operations ingests that data and generates insights via dashboards, charts, and alerts. Cloud Monitoring alerting helps you collaborate by integrating with Slack, PagerDuty, and more.

  

> ####

Reviewer 2: Not sure if this is required.

Author: changed on the latest commits

> We will be looking at ingesting data from Catchpoint into Cloud Monitoring.

Reviewer 2: Should we mention if we will be using REST API or webhook approach in this integration, also give link to Catchpoint KB like you have given for Google Cloud.

Author: Added telling its webhook.
Still waiting for kb to go live for Catchpoint KB

[https://cloud.google.com/monitoring](https://cloud.google.com/monitoring)

  

> ## **Installation &amp; Configuration**

Reviewer 2: We can remove the '**' here and in later heading as well.

Author: changed on the latest commits

>  **Google cloud Setup**:

Reviewer 2: Rather than '**' should we go for '###'?

Author: changed on the latest commits

>  **Enabling the Monitoring API**

Reviewer 2: If this is subsection to the above heading then '####'?

Author: changed on the latest commits

  

1. Select the [project](https://console.cloud.google.com/apis/dashboard) you will use to access the API.

> 2.  Click the Enable APIs and Service button.

Reviewer 2: All the links and button interaction should in inside single tick `, I believe "Enable APIs and Service" is one button, so it will be `Enable APIs and Service`. This way we can easily find important interactions on the page.

Reviewer 2: `Enable APIs and Service`

Reviewer 2: We can do the above for the entire documentation.

3. Search for &quot;Stackdriver&quot;.

4. In the search results, click through to &quot;Cloud Monitoring API&quot;.

5. If &quot;API enabled&quot; is displayed, then the API is already enabled. If not, click the Enable button.

  

> **Install cloud SDK on your local machine**

Reviewer 2: we can see if this can use heading formatting.

Author: changed on the latest commits

  

1. Download the [Cloud SDK installer.](https://dl.google.com/dl/cloudsdk/channels/rapid/GoogleCloudSDKInstaller.exe)

2. Launch the installer and follow the prompts. The installer is signed by Google LLC. Cloud SDK requires Python. Supported versions are 3.5 to 3.7, and 2.7.9 or higher. The installer will install all necessary dependencies, including the needed Python version.

3. After installation has completed, accept the following options:

  

> - Start Cloud SDK Shell

Reviewer 2: lets indent this with a tab and number as sub section rather then bullets.

Author: made changes for this

> The installer starts a terminal window and runs the gcloud init command.

Reviewer 2: Should we put this as a note?

Author: Added as note

> **Enabling cloud functions.**

Reviewer 2: heading?

Author: changed on the latest commits

  

1. Go to the project selector page

2. Make sure that billing is enabled for your Google Cloud project. Learn how to confirm billing is enabled for your project.

3. Enable the [Cloud Functions and Cloud Pub/Sub APIs.](https://console.cloud.google.com/flows/enableapi?apiid=cloudfunctions,pubsub&redirect=https://cloud.google.com/functions/docs/tutorials/pubsub)

4. Install and initialize the Cloud SDK.

5. Update gcloud components:
 ```bash
gcloud components update
```

> ##

Reviewer 2: is this required? may be a blank line?

Author: changed on the latest commits

  

**Clone the Repository to your local machine**

  

The repository has all the required NodeJS scripts to deploy webhooks.
Clone this repository to your local machine.

Navigate to the directory where the files were cloned and update project Id under .env file

   ```bash
cd <path to extracted directory>;

> ```

Reviewer 2: Convert this to step, with numbering.

Author: added numbering

  
  
 **Deploying cloud functions.**

  

Index.js has two functions called catchpointPublish and catchpointSubscribe.

Open Google Cloud SDK Shell and navigate to the directory where the NodeJS scripts was extracted.

 ```bash
cd <path to extracted directory>;

> ```

Reviewer 2: This can have steps with numbering.

Reviewer 2: do the same for later section if required.

  

1.  **Deploy publish function:**

  
 ```bash
gcloud functions deploy catchpointPublish --trigger-http --runtime nodejs10 --timeout=180 --trigger-http --allow-unauthenticated
```
Copy the URL once the deployment is successful. This will be webhook URL which will be added in Catchpoint portal.

  

1.  **Deploy Subscribe function:**

  
 ```bash
gcloud functions deploy catchpointSubscribe --trigger-topic catchpoint-webhook --timeout=180 --runtime nodejs10 --allow-unauthenticated

>   ```

Reviewer 2: I didnt understand the numbering here, also can be steps.

**Catchpoint Setup:**

  

Add the copied URL to Catchpoint ->Settings ->API page under Test Data Webhook.

**Note: Test data webhook should be enabled under the test properties page.**

## **Results**

  

To use Cloud Monitoring, you must have a Google Cloud project with billing enabled. The project must also be associated with a Workspace. Cloud Monitoring uses Workspaces to organize monitored Google Cloud projects.

> In the Google Cloud Console, go to Monitoring -\&gt; Overview this will create a workspace for you automatically for the first time.

Reviewer 2: this can be numbered.

Author: numbered

  
To view the metrics for a monitored resource using Metrics Explorer, do the following:

 1. From the Google Cloud Console, go to Monitoring. [https://console.cloud.google.com/monitoring](https://console.cloud.google.com/monitoring)
 2. In the Monitoring navigation pane, click Metrics Explorer.
 3. Enter the monitored resource name in the Find resource type and metric text box.
Resource type -> global. 
All the Catchpoint specific metrics will have id’s in the following format. custom.googleapis.com/global/catchpoint_Connect
custom.googleapis.com/global/catchpoint_Dns
custom.googleapis.com/global/catchpoint_Load
 4. Metrics explorer also allows to filter data points using node name or test id.
 5. Add all the required metrics and this save the chart to a dashboard.
[https://cloud.google.com/monitoring/charts/metrics-explorer
](https://cloud.google.com/monitoring/charts/metrics-explorer
)

>  6. Navigate to Monitoring->Dashboards to check out the metrics.

Reviewer 2: We can have the 1. approach for all sections.

Reviewer 2: add sub sections if required.
```
1. 
1. 
      1.
      1. 
```
